### PR TITLE
DNNL-BYOC enhancement

### PR DESCRIPTION
**Pull request by crazydemo**
- add more patterns in `dnnl.py`.
- add test cases, which have covered the support ops and patterns, in `test_dnnl.py`.
- add implementation of new added ops and patterns in `dnnl_json_runtime.cc`.
- add support to `dilated conv2d` with dnnl backend.

### CODE REVIEW

File: tests/python/relay/test_pass_partition_graph.py [hunk 1]
```diff
@@ -989,7 +989,7 @@ def test_partition():
         # conv + relu, conv + relu -> no fusion, 4 partition each with a single op
         test_detect_pattern([conv2d_bias_relu_pat], False, False, 4)
         # conv + bn + sigmoid + relu, conv + sigmoid + relu -> no fusion
-        test_detect_pattern([conv2d_bias_relu_pat, conv2d_relu_pat], True, True, 5)
+        test_detect_pattern([conv2d_bias_relu_pat, conv2d_relu_pat], True, True, 7)
```
Comment (jwfromm) [inline tests/python/relay/test_pass_partition_graph.py:992]: Would it make sense to add a few tests here? You've done a great job with tests that confirm the correctness of graph partitioned for DNNL, but havent tested that the new patterns are properly fused. All your new tests would pass even if none of the new pattern matches were offloaded as far as I can tell.

**Review by jwfromm — COMMENTED** [review]
Thanks for this change, it looks really excellent. I have a few small comments but am curious if youve done any benchmarking with this new pattern fusion. If so, how much do these changes speed up models you tried?

**Comment by crazydemo** [comment]
> Thanks for this change, it looks really excellent. I have a few small comments but am curious if youve done any benchmarking with this new pattern fusion. If so, how much do these changes speed up models you tried?

Thank you for your comments. We do have done some benchmark, and the related results are posted on [RFC](https://discuss.tvm.apache.org/t/rfc-byoc-intel-r-onednn-integration/11582).

Note that, the results in `RFC`  are collected with the `registered pattern` and optimization from `alter_op_layout pass`.
We are going to submit another two `layout transform` related PRs to fully support the performance gain claimed in `RFC`.

File: tests/python/contrib/test_dnnl.py [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+            result_key = mode + ("_dnnl" if use_dnnl else "")
+            if use_dnnl:
+                mod = dnnl.partition_for_dnnl(mod, params)
+                check_dnnl_used(mod)
+                with tvm.transform.PassContext(opt_level=3):
+                    func = relay.create_executor(
+                        mode, mod=mod, device=dev, target=target
+                    ).evaluate()
+            else:
+                with tvm.transform.PassContext(opt_level=3):
```
[the remark this answers is not in the document]
Comment (crazydemo) [reply tests/python/contrib/test_dnnl.py:74]: Fixed.

File: tests/python/relay/test_pass_partition_graph.py (hunk 1 above)
Comment (crazydemo) [reply tests/python/relay/test_pass_partition_graph.py:992]: Sure, more tests are added.

**Review by masahi — APPROVED** [review]

**Pull request merged**